### PR TITLE
Remove redundant `convert` in `_setindex!`

### PR DESCRIPTION
**Pull request by giordano**
Follow up to #56034, ref: https://github.com/JuliaLang/julia/pull/56034#discussion_r1798573573.

### CODE REVIEW

**Review by topolarity — APPROVED** [review]
Ah thanks @giordano !

I saw your comment and just hadn't had a chance to open the PR yet

**Comment by oscardssmith** [comment]
Can you also change the `Memory` `getindex` and `setindex!` to do this also?

**Comment by giordano** [comment]
Done, but I assumed you meant `setindex!` only, since `getindex` doesn't need to do conversions like that.

File: base/genericmemory.jl [hunk 1]
```diff
@@ -235,13 +235,17 @@ getindex(A::Memory, c::Colon) = copy(A)
...
 
-function setindex!(A::Memory{T}, x, i1::Int) where {T}
-    val = x isa T ? x : convert(T,x)::T
+function _setindex!(A::Memory{T}, x::T, i1::Int) where {T}
     ref = memoryrefnew(memoryref(A), i1, @_boundscheck)
     memoryrefset!(ref, val, :not_atomic, @_boundscheck)
     return A
 end
 
+function setindex!(A::Memory{T}, x, i1::Int) where {T}
```
Comment (topolarity) [inline base/genericmemory.jl:244]:
```suggestion
function setindex!(A::Memory{T}, x, i1::Int) where {T}
    @_propagate_inbounds_meta
```

**Comment by topolarity** [comment]
The atomics + `fill!` all have converts that could be lifted as well, if you want to take on Memory

File: base/genericmemory.jl (hunk 1 above)
Comment (giordano) [reply base/genericmemory.jl:244]: I think this is addressed by #56151 already.

Comment (giordano) [reply base/genericmemory.jl:244]: Wait, no, that's a different method!

**Review by oscardssmith — APPROVED** [review]

**Pull request merged**